src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -625,48 +625,67 @@ private static string Encode5987(StringSegment input)
             ? stackalloc byte[MaxStackAllocSizeBytes]
             : bufferFromPool = ArrayPool<byte>.Shared.Rent(maxInputBytes);
 
+        char[]? charBufferFromPool = null;
+        Span<char> tempCharBuffer = maxInputBytes <= MaxStackAllocSizeBytes
+            ? stackalloc char[MaxStackAllocSizeBytes]
+            : charBufferFromPool = ArrayPool<char>.Shared.Rent(maxInputBytes);
+
         var bytesWritten = Encoding.UTF8.GetBytes(input, inputBytes);
         inputBytes = inputBytes[..bytesWritten];
 
-        int totalBytesConsumed = 0;
-        while (totalBytesConsumed < inputBytes.Length)
+        while (inputBytes.Length > 0)
         {
-            if (Ascii.IsValid(inputBytes[totalBytesConsumed]))
+            var length = inputBytes.IndexOfAnyExcept(Rfc5987AttrChar);
+            if (length < 0)
+            {
+                length = inputBytes.Length;
+            }
+            int asciiCharCount = Encoding.ASCII.GetChars(inputBytes.Slice(0, length), tempCharBuffer);
+            builder.Append(tempCharBuffer[..asciiCharCount]);
+
+            inputBytes = inputBytes.Slice(length);
+            if (inputBytes.IsEmpty)
             {
-                // This is an ASCII char. Let's handle it ourselves.
+                break;
+            }
 
-                char c = (char)inputBytes[totalBytesConsumed];
-                if (!AttrChar.Contains(c))
+            length = inputBytes.IndexOfAny(Rfc5987AttrChar);
+            if (length < 0)
+            {
+                length = inputBytes.Length;
+            }
+
+            var toHexEscape = inputBytes.Slice(0, length);
+            while (toHexEscape.Length > 0)
+            {
+                if (Ascii.IsValid(toHexEscape[0]))
                 {
-                    HexEscape(builder, c);
+                    HexEscape(builder, toHexEscape[0]);
+                    toHexEscape = toHexEscape.Slice(1);
                 }
                 else
                 {
-                    builder.Append(c);
+                    Rune.DecodeFromUtf8(toHexEscape, out Rune r, out int bytesConsumedForRune);
```

Comment:
You didn't change this and it probably isn't worth changing, but I kind of wonder whether it would simplify the code (without hurting perf?) to just escape a single non-ascii character per iteration of the outer loop (rather than computing the length of the non-ascii sequence and processing it in multiple nested loops).

Reply:
Yes, I don't think we need this `if` at all. It isn't there for the runtime implementation either, and Rune is only to tell the bytes count, but the old implementation also Hex's a single byte only independent of whatever Rune says.

Reply:
updated